File: frontend/src/api/reading.ts
/** 阅读侧 API：书架 / 书目录 / 文章 / 全文搜索（双模式） */
import { detectMode, getJSON, readUrl, type IndexPayload } from './http'
import type { ArticlePayload, BookPayload, SearchHit } from './types'

interface StaticSearchItem {
  pid: string
  bid: string
  slug: string
  t: string
  x: string
}

let staticSearchIndex: StaticSearchItem[] | null = null

function esc(s: string): string {
  return s.replace(/[&<>"]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' })[c] as string)
}

export async function getIndex(): Promise<IndexPayload> {
  if ((await detectMode()) === 'api') return getJSON<IndexPayload>('api/index')
  return getJSON<IndexPayload>('d/index.json')
}

export async function getBook(pid: string, bid: string): Promise<BookPayload> {
  if ((await detectMode()) === 'api') return getJSON<BookPayload>(`api/book/${pid}/${bid}`)
  return getJSON<BookPayload>(`d/${pid}/${bid}/toc.json`)
}

export async function getArticle(pid: string, bid: string, slug: string): Promise<ArticlePayload> {
  const path = slug.split('/').map(encodeURIComponent).join('/')
  if ((await detectMode()) === 'api') return getJSON<ArticlePayload>(`api/article/${pid}/${bid}/${path}`)
  return getJSON<ArticlePayload>(`d/${pid}/${bid}/${path}.json`)
}

export async function searchAll(q: string, pid = ''): Promise<SearchHit[]> {
  const query = q.trim()
  if (!query) return []
  if ((await detectMode()) === 'api') {
    const r = await getJSON<{ items: SearchHit[] }>(
      `api/search?q=${encodeURIComponent(query)}&pid=${encodeURIComponent(pid)}`)
    return r.items
  }
  // 离线静态模式：本地过滤预构建索引
  if (!staticSearchIndex) {
    staticSearchIndex = await getJSON<StaticSearchItem[]>('d/search.json')
  }
  const t = query.toLowerCase()
  return staticSearchIndex
    .filter((a) => (!pid || a.pid === pid) && (a.t + ' ' + a.x).toLowerCase().includes(t))
    .slice(0, 60)
    .map((a) => {
      const low = a.x.toLowerCase()
      const i = low.indexOf(t)
      const win = i < 0 ? a.x.slice(0, 90) : a.x.slice(Math.max(0, i - 24), i + 90)
      const re = new RegExp(query.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi')
      const marked = esc(win).replace(re, (m) => `<mark>${m}</mark>`)
      return { pid: a.pid, bid: a.bid, slug: a.slug, title: a.t, snip: marked }
    })
}

export function hitUrl(hit: SearchHit): string {
  return readUrl(hit.pid, hit.bid, hit.slug)
}

File: backend/src/api/routes/reading.py
"""阅读接口（控制器层）：书架 / 书目录 / 文章 / 全文搜索。"""
from fastapi import APIRouter, HTTPException, Query
from ...models import repository
from ...services import content_service, note_service

router = APIRouter(prefix='/api', tags=['reading'])


@router.get('/index')
def index():
    return repository.index_payload()


@router.get('/book/{pid}/{bid}')
def book(pid: str, bid: str):
    payload = content_service.book_payload(pid, bid)
    if not payload:
        raise HTTPException(status_code=404, detail='book not found')
    return payload


@router.get('/article/{pid}/{bid}/{slug:path}')
def article(pid: str, bid: str, slug: str):
    payload = content_service.render_article(pid, bid, slug)
    if not payload:
        raise HTTPException(status_code=404, detail='article not found')
    return payload


@router.get('/search')
def search(q: str = Query(''), pid: str = Query(''), limit: int = Query(60)):
    items = repository.search(q, pid=pid, limit=limit)
    for r in items:
        r['kind'] = 'article'
    if not pid:
        items += note_service.search(q, limit=max(5, limit // 6))
    return {'items': items}
